### DISCUSSION TITLE
Compatibility with multi-controlnet batch

### DISCUSSION THREAD

**Discussion by ljleb** [pkuliyi2015/multidiffusion-upscaler-for-automatic1111#89] [General]
Hi, while implementing https://github.com/Mikubill/sd-webui-controlnet/pull/683 I realized this extension does not work with the solution I found. I would appreciate a bit of help regarding how multidiffusion interacts with the controlnet extension for a smoother upgrade.

If it is possible to update the code here to depend less on the internals of controlnet, maybe we could look into that? There may be other extensions doing the same thing so I am not sure what the best approach is.

**Comment by ljleb** [comment]
It seems the control maps are cached and tiled once at the beginning of the sampling process. This will not work as batch processing needs to switch the control maps during processing.

My solution currently needs to know the effective sampling steps that happen during normal processing + hires processing, to be able to swap the control maps at the right moment. As multidiffusion changes these amounts arbitrarily, control maps are being swapped repeatedly during each batch image.

**Reply by pkuliyi2015** [reply]
Hello, the contronet tensor cache can be easily refreshed when neccessary by calling init_controlnet in abstractdiffusion.py. I just want to know when do I need to refresh that? So everything will be compatible.

For example, just refresh when the sampler is created, is this OK? I will push an update immediately for your test.

**Reply by LightChaser666** [reply]
Compatibility test has been pushed into dev/seed branch.

**Reply by ljleb** [reply]
Hi, thanks for the help!
Is the sampler created for each iteration of a batch? (i.e. `batch_count` in the code)
If so, then it could be part of the solution I think. The current batch implementation duplicates the prompts in `p` as many times as there are controlnet batches to do, which means `script.process` is called a single time:

https://github.com/Mikubill/sd-webui-controlnet/blob/6ab382c77e9b5926fdfaae00e9e1ffe57c85d844/scripts/controlnet.py#L817-L821

Note that there is another issue, which is that the sampler steps used to switch the control maps do not align anymore with the effective steps used during sampling:

https://github.com/Mikubill/sd-webui-controlnet/blob/6ab382c77e9b5926fdfaae00e9e1ffe57c85d844/scripts/hook.py#L127-L134

I'll try to play with the code and see if I find a simple fix.

**Reply by ljleb** [reply]
Ah the github thread wasn't updated for me. Thanks for the quick fix, I'll see if I can make it work with the dev branch.

**Reply by pkuliyi2015** [reply]
Yes the sampler is not reused normally. When you use batch_count > 1 or batch processing, then every time the sampler will be recreated. So I hooked into the sd_samplers.create_sampler(), but I changed the recreation behavior by returning a previous one because recreating a sampler is costly and unnecessary. 

I believe if I refresh control tensors when the sampler is created, the incompatibility with your PR will be resolved; Unless you haven't got your tensor prepared before sampling start (i.e., at the time when sample_img2img is called).

**Reply by ljleb** [reply]
Thanks a lot, it seems to be working. Using `script.process_batch` to switch the control maps fixed my second issue also.

However it seems that the first image of the batch isn't tiled, even though the rest of the batch is. (so only index 0 is not tiled) I tried putting a breakpoint at the new location where `init_controlnet` is called (branch dev/seed) and it seems that it is not called for the first image? As I have not modified the code of multidiffusion, you may want to take a look at this.

**Reply by pkuliyi2015** [reply]
Fixed with thanks.
